File: src/main/java/com/erzbir/mirai/numeron/plugins/filesaver/FileSaver.java
package com.erzbir.mirai.numeron.plugins.filesaver;

import com.erzbir.mirai.numeron.configs.entity.WhiteList;
import com.erzbir.mirai.numeron.plugins.Plugin;
import com.erzbir.mirai.numeron.plugins.PluginRegister;
import com.erzbir.mirai.numeron.processor.Command;
import com.erzbir.mirai.numeron.store.DefaultStore;
import com.erzbir.mirai.numeron.store.RedisStore;
import com.erzbir.mirai.numeron.utils.NetUtil;
import net.mamoe.mirai.Bot;
import net.mamoe.mirai.contact.Contact;
import net.mamoe.mirai.contact.Group;
import net.mamoe.mirai.contact.file.AbsoluteFile;
import net.mamoe.mirai.contact.file.AbsoluteFileFolder;
import net.mamoe.mirai.event.EventChannel;
import net.mamoe.mirai.event.events.BotEvent;
import net.mamoe.mirai.event.events.MessageEvent;
import net.mamoe.mirai.message.data.*;
import net.mamoe.mirai.utils.ExternalResource;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Arrays;
import java.util.Date;
import java.util.Locale;
import java.util.Objects;

/**
 * @author Erzbir
 * @Date: 2022/12/12 20:41
 * 由群友提供kt代码, 我改为java代码
 */
@Plugin
public class FileSaver implements PluginRegister {
    private final String mainDir = "erzbirnumeron/";
    private final String botDir = mainDir + "bots/";
    private final Path storeLocation = Path.of(botDir, "qq_files");
    private final Path picStoreLocation = Path.of(storeLocation.toAbsolutePath().toString(), "pic");
    private final Path fileStoreLocation = Path.of(storeLocation.toAbsolutePath().toString(), "file");
    private final Path audioStoreLocation = Path.of(storeLocation.toAbsolutePath().toString(), "audio");
    private boolean autoPick = false;

    {
        try {

            if (!Files.exists(storeLocation)) {
                Files.createDirectory(storeLocation);
            }
            if (!Files.exists(picStoreLocation)) {
                Files.createDirectory(picStoreLocation);
            }
            if (!Files.exists(fileStoreLocation)) {
                Files.createDirectory(fileStoreLocation);
            }
            if (!Files.exists(audioStoreLocation)) {
                Files.createDirectory(audioStoreLocation);
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    @Command(name = "文件保存和发送", dec = "图片/文件/文本", help = "/fput  /fget  /dl")
    @Override
    public void register(Bot bot, EventChannel<BotEvent> channel) {
        channel.subscribeAlways(MessageEvent.class, it -> {
            int[] ids = it.getSource().getIds();
            MessageChain message = it.getMessage();
            DefaultStore defaultStore = DefaultStore.getInstance();
            RedisStore redisStore = RedisStore.getInstance();
            defaultStore.save(ids[0], message);
            long sender = it.getSender().getId();
            String content = message.contentToString();
            Contact contact = it.getSubject();
            if (autoPick) {
                for (Message m : message) {
                    if (m instanceof Image pic) {
                        String picUrl = Image.queryUrl(pic);
                        try {
                            NetUtil.downloadTo(picUrl, new File(picStoreLocation.toString(), pic.getImageId()));
                            redisStore.set("pic_" + pic.getImageId(), pic.getImageId(), -1L);
                        } catch (Exception e) {
                            e.printStackTrace();
                            return;
                        }
                    } else if (m instanceof PlainText plain) {
                        try {
                            redisStore.set("plain_" + new Date().getTime(), plain.contentToString(), -1L);
                        } catch (Exception e) {
                            return;
                        }
                    } else if (m instanceof FileMessage file) {
                        if (contact instanceof Group group) {
                            AbsoluteFile file2 = file.toAbsoluteFile(group);
                            if (file2 == null) {
                                return;
                            }
                            NetUtil.downloadTo(file2.getUrl(), new File(fileStoreLocation.toString(), file2.getName()));
                            redisStore.set("file_" + file2.getId(), file2.getName(), -1L);
                        }
                    }
                }
            }
            if (content.startsWith("/fput")) {
                if (!WhiteList.INSTANCE.contains(sender)) {
                    return;
                }
                String[] command = content.split("\\s+?");
                if (command.length != 2) {
                    contact.sendMessage("Usage: /fput [filename]");
                    return;
                }
                int[] ids1 = Objects.requireNonNull(message.get(QuoteReply.Key)).getSource().getIds();
                MessageChain find = defaultStore.find(ids1[0]);
                if (find == null) {
                    contact.sendMessage("你选中的消息无法找到");
                    return;
                }
                boolean saved = false;
                for (Message ele : find) {
                    if (ele instanceof Image image) {
                        String picUrl = Image.queryUrl(image);
                        try {
                            NetUtil.downloadTo(picUrl, new File(picStoreLocation.toString(), command[1] + "." + image.getImageType().name()));
                            redisStore.set("pic_" + image.getImageId(), command[1] + "." + image.getImageType().name().toLowerCase(Locale.getDefault()), -1L);
                            saved = true;
                        } catch (Exception e) {
                            contact.sendMessage(
                                    new PlainText("图片保存失败").plus(
                                            new PlainText("\n").plus(
                                                    new PlainText(e.getMessage()))));
                            return;
                        }
                        contact.sendMessage("保存图片成功");
                    }
                    if (ele instanceof PlainText plain) {
                        try {
                            redisStore.set("plain_" + new Date().getTime(), plain.contentToString(), -1L);
                        } catch (Exception e) {
                            contact.sendMessage(
                                    new PlainText("文本保存失败").plus(
                                            new PlainText("\n").plus(
                                                    new PlainText(e.getMessage()))));
                            return;
                        }
                        contact.sendMessage("保存文本成功");
                    }
                    if (ele instanceof Audio audio) {
                        return;
                    }
                    if (ele instanceof FileMessage file) {
                        if (contact instanceof Group group) {
                            AbsoluteFile file2 = file.toAbsoluteFile(group);
                            try {
                                if (file2 == null) {
                                    return;
                                }
                                NetUtil.downloadTo(
                                        file2.getUrl(),
                                        new File(fileStoreLocation.toString(), command[1] + "." + AbsoluteFileFolder.Companion.getExtension(file2)));
                                redisStore.set("file_" + file.getId(),
                                        command[1] + "." + AbsoluteFileFolder.Companion.getExtension(file2).toLowerCase(Locale.getDefault()), -1L);
                                saved = true;
                            } catch (Exception e) {
                                contact.sendMessage(new PlainText("文件保存失败")
                                        .plus(new PlainText("\n")
                                                .plus(new PlainText(e.getMessage()))));
                                return;
                            }
                            contact.sendMessage("保存文件成功");
                        }

                    }
                    if (!saved) {
                        contact.sendMessage("你选中的消息没有可以保存的内容");
                    }
                }
            }


            if (content.startsWith("/fget")) {
                if (!WhiteList.INSTANCE.contains(sender)) {
                    return;
                }
                String[] command = content.split("\\s+");
                if (command.length != 3 && command.length != 5) {
                    contact.sendMessage("Usage: /fget [pic|file] [filename] or /get [pic|file] [filename] [width] [height]");
                    return;
                }
                switch (command[1]) {
                    case "pic" -> {
                        if (Files.exists(Path.of(picStoreLocation.toString(), command[2]))) {
                            try {
                                Image uploadImage = Contact.uploadImage(contact, new File(Path.of(picStoreLocation.toString(), command[2]).toString()));
                                if (command.length == 5) {
                                    Image.Builder newBuilder = Image.Builder.newBuilder(uploadImage.getImageId());
                                    newBuilder.setWidth(Integer.parseInt(command[3]));
                                    newBuilder.setHeight(Integer.parseInt(command[4]));
                                    contact.sendMessage(newBuilder.build());
                                    return;
                                }
                                contact.sendMessage(uploadImage);
                            } catch (Exception e) {
                                contact.sendMessage(new PlainText("文件发送失败")
                                        .plus(new PlainText("\n")
                                                .plus(new PlainText(e.getMessage()))));
                                return;
                            }
                        } else {
                            contact.sendMessage("没有这张图片");
                        }
                    }

                    case "file" -> {
                        if (Files.exists(Path.of(fileStoreLocation.toString(), command[2]))) {
                            if (contact instanceof Group group) {
                                try {
                                    ExternalResource toExternalResource = ExternalResource.create(new File(Path.of(fileStoreLocation.toString(), command[2]).toAbsolutePath().toString()));
                                    group.getFiles().uploadNewFile(command[2], toExternalResource);
                                } catch (Exception e) {
                                    contact.sendMessage(
                                            new PlainText("文件发送失败")
                                                    .plus(new PlainText("\n")
                                                            .plus(new PlainText(e.getMessage()))));
                                    return;
                                }
                            } else {
                                contact.sendMessage("只支持向群聊发送文件");
                            }
                        } else {
                            contact.sendMessage("没有这个文件");
                        }
                    }
                }
            }

            if (content.startsWith("/fls")) {
                if (!WhiteList.INSTANCE.contains(sender)) {
                    return;
                }
                String[] command = content.split("\\s+?");
                if (command.length != 2) {
                    contact.sendMessage("Usage: /fls [pic|file|plain]");
                    return;
                }
                switch (command[1]) {
                    case "pic" -> {
                        String[] f = picStoreLocation.toFile().list();
                        if (f == null || f.length == 0) {
                            contact.sendMessage("暂无文件");
                        } else {
                            contact.sendMessage(Arrays.toString(f).replace(",", "\n"));
                        }
                    }

                    case "file" -> {
                        String[] f = fileStoreLocation.toFile().list();
                        if (f == null || f.length == 0) {
                            contact.sendMessage("暂无文件");
                        } else {
                            contact.sendMessage(Arrays.toString(f).replace(",", "\n"));
                        }
                    }

                    case "plain" -> {
                        String m = redisStore.getPlain().toString().replaceAll(",", "\n");
                        if (m.isEmpty()) {
                            contact.sendMessage("暂无文本");
                        } else {
                            contact.sendMessage(m);
                        }
                    }
                }
            }

            if (content.startsWith("/auto_pick")) {
                if (!WhiteList.INSTANCE.contains(sender)) {
                    return;
                }
                String[] command = content.split("\\s+?");
                if (command.length != 2) {
                    contact.sendMessage("Usage: /auto_pick [true|false]");
                    return;
                }
                switch (command[1]) {
                    case "true" -> {
                        autoPick = true;
                        contact.sendMessage("已开启自动保存");
                    }

                    case "false" -> {
                        autoPick = false;
                        contact.sendMessage("已关闭自动保存");
                    }
                }
            }

            if (content.startsWith("/dl")) {
                if (!WhiteList.INSTANCE.contains(sender)) {
                    return;
                }
                String[] command = content.split("\\s+?");
                if (command.length != 3) {
                    contact.sendMessage("Usage: /dl [url] [filename]");
                    return;
                }
                contact.sendMessage("正在下载文件...");
                try {
                    NetUtil.downloadTo(command[1], new File(fileStoreLocation.toString(), command[2]));
                } catch (Exception e) {
                    contact.sendMessage(new PlainText("文件下载失败")
                            .plus(new PlainText("\n")
                                    .plus(new PlainText(e.getMessage()))));
                    return;
                }
                redisStore.set("file_${command[2]}", command[2], -1L);
                contact.sendMessage("文件" + command[2] + "下载完成");
            }

            if (content.startsWith("/frm")) {
                if (!WhiteList.INSTANCE.contains(sender)) {
                    return;
                }
                String[] command = content.split("\\s+");
                if (command.length != 3) {
                    contact.sendMessage("Usage: /frm [pic|file|plain] [filename]");
                    return;
                }
                switch (command[1]) {
                    case "pic" -> {

                        try {
                            Files.delete(Path.of(picStoreLocation.toString(), command[2]));
                            redisStore.removePic(command[2]);
                        } catch (Exception e) {
                            contact.sendMessage("删除失败 " + e.getMessage());
                            return;
                        }
                        contact.sendMessage("删除图片 " + command[2] + " 成功");
                    }

                    case "file" -> {
                        try {
                            Files.delete(Path.of(fileStoreLocation.toString(), command[2]));
                            redisStore.removeFile(command[2]);
                        } catch (Exception e) {
                            contact.sendMessage(e.getMessage());
                            return;
                        }
                        contact.sendMessage("删除文件[${command[2]}]成功");
                    }

                    case "plain" -> {
                        try {
                            redisStore.removePlain(command[2]);
                        } catch (Exception e) {
                            contact.sendMessage(e.getMessage());
                        }
                        contact.sendMessage("删除文本 " + command[2] + " 成功");
                    }
                }
            }
        });
    }
}

File: src/main/java/com/erzbir/mirai/numeron/plugins/filesaver/FileSaver.kt
package com.erzbir.mirai.numeron.plugins.filesaver

import com.erzbir.mirai.numeron.configs.GlobalConfig
import com.erzbir.mirai.numeron.plugins.Plugin
import com.erzbir.mirai.numeron.plugins.PluginRegister
import com.erzbir.mirai.numeron.processor.Command
import com.erzbir.mirai.numeron.store.DefaultStore
import com.erzbir.mirai.numeron.store.RedisStore
import com.erzbir.mirai.numeron.utils.NetUtil
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import net.mamoe.mirai.Bot
import net.mamoe.mirai.contact.Contact.Companion.uploadImage
import net.mamoe.mirai.contact.Group
import net.mamoe.mirai.contact.file.AbsoluteFileFolder.Companion.extension
import net.mamoe.mirai.event.EventChannel
import net.mamoe.mirai.event.events.BotEvent
import net.mamoe.mirai.event.events.MessageEvent
import net.mamoe.mirai.message.data.*
import net.mamoe.mirai.message.data.Image.Key.queryUrl
import net.mamoe.mirai.utils.ExternalResource.Companion.toExternalResource
import java.io.File
import java.io.IOException
import java.nio.file.Files
import java.nio.file.Path
import java.util.*
import kotlin.io.path.absolutePathString

@Plugin
object FileSaver : PluginRegister {
    private var autoPick = false
    private val storeLocation = Path.of(GlobalConfig.botDir, "qq_files")
    private val picStoreLocation: Path = Path.of(storeLocation.absolutePathString(), "pic")
    private val fileStoreLocation: Path = Path.of(storeLocation.absolutePathString(), "file")
    private val audioStoreLocation: Path = Path.of(storeLocation.absolutePathString(), "audio")

    init {
        if (!Files.exists(storeLocation)) {
            Files.createDirectory(storeLocation)
        }
        if (!Files.exists(picStoreLocation)) {
            Files.createDirectory(picStoreLocation)
        }
        if (!Files.exists(fileStoreLocation)) {
            Files.createDirectory(fileStoreLocation)
        }
        if (!Files.exists(audioStoreLocation)) {
            Files.createDirectory(audioStoreLocation)
        }
    }

    @Command(name = "文件保存和发送", dec = "图片/文件/文本", help = "/fput  /fget  /dl")
    override fun register(bot: Bot, channel: EventChannel<BotEvent>) {
        channel.subscribeAlways<MessageEvent> { it ->
            val ids = it.source.ids
            val message1 = it.message
            val defaultStore = DefaultStore.getInstance()
            val redisStore = RedisStore.getInstance()
            defaultStore.save(ids[0], message1)
            val sender = it.sender.id
            val content = it.message.contentToString()
            if (autoPick) {
                for (m in message1) {
                    if (m is Image) {
                        val picUrl = m.queryUrl()
                        try {
                            NetUtil.downloadTo(picUrl, File(picStoreLocation.absolutePathString(), m.imageId))
                            redisStore.set("pic_" + m.imageId, m.imageId, -1L)
                        } catch (e: Exception) {
                            e.printStackTrace()
                            return@subscribeAlways
                        }
                    } else if (m is PlainText) {
                        try {
                            redisStore.set("plain_" + Date().time.toString(), m.content, -1L)
                        } catch (e: Exception) {
                            return@subscribeAlways
                        }
                    } else if (m is FileMessage) {
                        if (it.subject is Group) {
                            val file = m.toAbsoluteFile(it.subject as Group)
                            file?.let { f ->
                                try {
                                    NetUtil.downloadTo(
                                        f.getUrl()!!,
                                        File(fileStoreLocation.absolutePathString(), file.name)
                                    )
                                    redisStore.set("file_" + file.id, file.name, -1L)
                                } catch (e: Exception) {
                                    e.printStackTrace()
                                    return@subscribeAlways
                                }
                            }
                        }

                    }
                }
            }
            when {
                content.startsWith("/fput") -> {
                    if (!GlobalConfig.whiteList.contains(sender)) {
                        return@subscribeAlways
                    }
                    val command = content.split(" ").filter { it.isNotEmpty() }
                    if (command.size != 2) {
                        it.subject.sendMessage("Usage: /fput [filename]")
                        return@subscribeAlways
                    }
                    val ids1 = it.message[QuoteReply.Key]!!.source.ids
                    val find = defaultStore.find(ids1[0])
                    if (find == null) {
                        it.subject.sendMessage("你选中的消息无法找到")
                        return@subscribeAlways
                    }
                    println(find)
                    find.let { m ->
                        var saved = false
                        for (ele in m) {
                            if (ele is Image) {
                                val picUrl = ele.queryUrl()
                                try {
                                    NetUtil.downloadTo(
                                        picUrl,
                                        File(
                                            picStoreLocation.absolutePathString(),
                                            "${command[1]}.${ele.imageType.name}"
                                        )
                                    )
                                    redisStore.set(
                                        "pic_" + ele.imageId,
                                        "${command[1]}.${ele.imageType.name.lowercase(Locale.getDefault())}",
                                        -1L
                                    )
                                    saved = true
                                } catch (e: Exception) {
                                    it.subject.sendMessage(
                                        PlainText("图片保存失败").plus(
                                            PlainText("\n").plus(
                                                PlainText(
                                                    e.message ?: "未知错误"
                                                )
                                            )
                                        )
                                    )
                                    return@subscribeAlways
                                }
                                it.subject.sendMessage("保存图片成功")
                            }
                            if (ele is PlainText) {
                                try {
                                    redisStore.set("plain_" + Date().time.toString(), ele.content, -1L)
                                } catch (e: Exception) {
                                    it.subject.sendMessage(
                                        PlainText("文本保存失败").plus(
                                            PlainText("\n").plus(
                                                PlainText(
                                                    e.message ?: "未知错误"
                                                )
                                            )
                                        )
                                    )
                                    return@subscribeAlways
                                }
                                it.subject.sendMessage("保存文本成功")
                            }
                            if (ele is Audio) {
                                return@subscribeAlways
                            }
                            if (ele is FileMessage) {
                                if (it.subject is Group) {
                                    val file = ele.toAbsoluteFile(it.subject as Group)
                                    file?.let { f ->
                                        try {
                                            NetUtil.downloadTo(
                                                f.getUrl()!!,
                                                File(
                                                    fileStoreLocation.absolutePathString(),
                                                    "${command[1]}.${file.extension}"
                                                )
                                            )
                                            redisStore.set(
                                                "file_" + ele.id, "${command[1]}.${
                                                    file.extension.lowercase(
                                                        Locale.getDefault()
                                                    )
                                                }", -1L
                                            )
                                            saved = true
                                        } catch (e: Exception) {
                                            it.subject.sendMessage(
                                                PlainText("文件保存失败").plus(
                                                    PlainText("\n").plus(
                                                        PlainText(e.message ?: "未知错误")
                                                    )
                                                )
                                            )
                                            return@subscribeAlways
                                        }
                                        it.subject.sendMessage("保存文件成功")
                                    }
                                }

                            }
                        }
                        if (!saved) {
                            it.subject.sendMessage("你选中的消息没有可以保存的内容")
                        }
                    }
                }

                content.startsWith("/fget") -> {
                    if (!GlobalConfig.whiteList.contains(sender)) {
                        return@subscribeAlways
                    }
                    val command = content.split(" ").filter { it.isNotEmpty() }
                    if (command.size != 3 && command.size != 5) {
                        it.subject.sendMessage("Usage: /fget [pic|file] [filename] or /get [pic|file] [filename] [width] [height]")
                        return@subscribeAlways
                    }
                    when (command[1]) {
                        "pic" -> {
                            if (Files.exists(Path.of(picStoreLocation.absolutePathString(), command[2]))) {
                                try {
                                    val uploadImage = it.subject.uploadImage(
                                        File(
                                            Path.of(picStoreLocation.absolutePathString(), command[2])
                                                .absolutePathString()
                                        )
                                    )
                                    if (command.size == 5) {
                                        val newBuilder = Image.Builder.newBuilder(uploadImage.imageId)
                                        newBuilder.width = command[3].toInt()
                                        newBuilder.height = command[4].toInt()
                                        println("使用这个")
                                        it.subject.sendMessage(newBuilder.build())
                                        return@subscribeAlways
                                    }
                                    it.subject.sendMessage(uploadImage)
                                } catch (e: Exception) {
                                    it.subject.sendMessage(
                                        PlainText("文件发送失败").plus(
                                            PlainText("\n").plus(
                                                PlainText(
                                                    e.message ?: "未知错误"
                                                )
                                            )
                                        )
                                    )
                                    return@subscribeAlways
                                }
                            } else {
                                it.subject.sendMessage("没有这张图片")
                            }
                        }

                        "file" -> {
                            if (Files.exists(Path.of(fileStoreLocation.absolutePathString(), command[2]))) {
                                if (it.subject is Group) {
                                    try {
                                        val toExternalResource = File(
                                            Path.of(fileStoreLocation.absolutePathString(), command[2])
                                                .absolutePathString()
                                        ).toExternalResource()
                                        (it.subject as Group).files.uploadNewFile(command[2], toExternalResource)
                                    } catch (e: Exception) {
                                        it.subject.sendMessage(
                                            PlainText("文件发送失败").plus(
                                                PlainText("\n").plus(
                                                    PlainText(e.message ?: "未知错误")
                                                )
                                            )
                                        )
                                        return@subscribeAlways
                                    }
                                } else {
                                    it.subject.sendMessage("只支持向群聊发送文件")
                                }
                            } else {
                                it.subject.sendMessage("没有这个文件")
                            }
                        }
                    }
                }

                content.startsWith("/fls") -> {
                    if (!GlobalConfig.whiteList.contains(sender)) {
                        return@subscribeAlways
                    }
                    val command = content.split(" ").filter { it.isNotEmpty() }
                    if (command.size != 2) {
                        it.subject.sendMessage("Usage: /fls [pic|file|plain]")
                        return@subscribeAlways
                    }
                    when (command[1]) {
                        "pic" -> {
                            val f = picStoreLocation.toFile().list()
                            if (f == null || f.isEmpty()) {
                                it.subject.sendMessage("暂无文件")
                            } else {
                                it.subject.sendMessage(Arrays.toString(f).replace(",", "\n"))
                            }
                        }

                        "file" -> {
                            val f = fileStoreLocation.toFile().list()
                            if (f == null || f.isEmpty()) {
                                it.subject.sendMessage("暂无文件")
                            } else {
                                it.subject.sendMessage(Arrays.toString(f).replace(",", "\n"))
                            }
                        }

                        "plain" -> {
                            val m = redisStore.plain.joinToString("\n")
                            if (m.isEmpty()) {
                                it.subject.sendMessage("暂无文本")
                            } else {
                                it.subject.sendMessage(m)
                            }
                        }
                    }
                }

                content.startsWith("/auto_pick") -> {
                    if (!GlobalConfig.whiteList.contains(sender)) {
                        return@subscribeAlways
                    }
                    val command = content.split(" ").filter { it.isNotEmpty() }
                    if (command.size != 2) {
                        it.subject.sendMessage("Usage: /auto_pick [true|false]")
                        return@subscribeAlways
                    }
                    when (command[1]) {
                        "true" -> {
                            autoPick = true
                            it.subject.sendMessage("已开启自动保存")
                        }

                        "false" -> {
                            autoPick = false
                            it.subject.sendMessage("已关闭自动保存")
                        }
                    }
                }

                content.startsWith("/dl") -> {
                    if (!GlobalConfig.whiteList.contains(sender)) {
                        return@subscribeAlways
                    }
                    val command = content.split(" ").filter { it.isNotEmpty() }
                    if (command.size != 3) {
                        it.subject.sendMessage("Usage: /dl [url] [filename]")
                        return@subscribeAlways
                    }
                    it.subject.sendMessage("正在下载文件...")
                    try {
                        NetUtil.downloadTo(command[1], File(fileStoreLocation.absolutePathString(), command[2]))
                    } catch (e: Exception) {
                        it.subject.sendMessage(
                            PlainText("文件下载失败").plus(
                                PlainText("\n").plus(
                                    PlainText(
                                        e.message ?: "未知错误"
                                    )
                                )
                            )
                        )
                        return@subscribeAlways
                    }
                    redisStore.set("file_${command[2]}", command[2], -1L)
                    it.subject.sendMessage("文件${command[2]}下载完成")
                }

                content.startsWith("/frm") -> {
                    if (!GlobalConfig.whiteList.contains(sender)) {
                        return@subscribeAlways
                    }
                    val command = content.split(" ")
                    if (command.size != 3) {
                        it.subject.sendMessage("Usage: /frm [pic|file|plain] [filename]")
                        return@subscribeAlways
                    }
                    when (command[1]) {
                        "pic" -> {
                            withContext(Dispatchers.IO) {
                                try {
                                    Files.delete(Path.of(picStoreLocation.absolutePathString(), command[2]))
                                } catch (e: IOException) {
                                    it.subject.sendMessage("删除失败 ${e.message}")
                                    return@withContext
                                }
                            }
                            redisStore.removePic(command[2])
                            it.subject.sendMessage("删除图片[${command[2]}]成功")
                        }

                        "file" -> {
                            withContext(Dispatchers.IO) {
                                try {
                                    Files.delete(Path.of(fileStoreLocation.absolutePathString(), command[2]))
                                } catch (e: IOException) {
                                    it.subject.sendMessage("删除失败 ${e.message}")
                                    return@withContext
                                }
                            }
                            redisStore.removeFile(command[2])
                            it.subject.sendMessage("删除文件[${command[2]}]成功")
                        }

                        "plain" -> {
                            redisStore.removePlain(command[2])
                            it.subject.sendMessage("删除文本[${command[2]}]成功")
                        }
                    }
                }
            }
        }
    }
}
